### ISSUE TITLE
about create application

### ISSUE THREAD

**Issue by upccup** [gambol99/go-marathon#210]
the create applicaition function is 
```
func (r *marathonClient) CreateApplication(application *Application) (*Application, error) {
	result := new(Application)
	if err := r.apiPost(marathonAPIApps, &application, result); err != nil {
		return nil, err
	}

	return result, nil
}
```

and my question is why the parameter need a pointer, and if is a pointer why 
```
if err := r.apiPost(marathonAPIApps, &application, result); err != nil {
```
use & and if parameter use pointer why create a new struct as result and not use the pointer of the parameter

**Comment by timoreimann** [comment]
@upccup Thanks for the feedback. I'll try to answer your questions.

My guess is that the `application` parameter passed to `CreateApplication()` is a pointer because the struct is fairly big, and passing a pointer is more efficient as no fields will need to be copied. (Admittedly, the vast majority of them are pointers, so with a non-pointer argument we'd be copying mostly memory addresses. Still, that's a fair amount of data given the size of the struct.) This seems to be in line with the [Go code review advice on passing values](https://github.com/golang/go/wiki/CodeReviewComments#pass-values).

The `&` operator should be omitted. I was curious why this hasn't bitten us yet. Turns out that `json.Marshal` (which eventually runs on the parameter) recursively unwraps pointer types until it either hits a `nil` or an underlying type to parse. So the pointer-to-pointer-to-`Application` becomes a pointer-to-`Application` during serialization, which is what we should have used in the first place.
Happy to review any PR you may want to submit. Otherwise, I can fix this once I find some free time. (We do make the same mistake at multiple calls to `r.api*`.)

As to your last question: I think it's just cleaner to separate the input from the output parameter and not conflate the two. For example, the returned application instance may have some default values set which you did not set on the input app instance. That could be a source of confusion if you wanted to reuse your input variable on the client end. Another argument is that in the case of an error, there's no guarantee what your variable may look like on return of the function: go-marathon may have populated the struct halve way through when it encountered an erroneous situation. I don't think we do something like this yet but it's theoretically possible for some point in the future.
In summary, not being destructive on the input variables should minimize you number of surprises you may run into.
